flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1573,7 +1577,19 @@ public void run() {
                 // the user-defined cancel method may throw errors.
                 // we need do continue despite that
                 try {
-                    invokable.cancel();
+                    Future<Void> cancellationFuture = invokable.cancel();
+                    // Wait for any active actions to complete (e.g. timers, mailbox actions)
+                    // Before that, interrupt to notify them about cancellation
+                    if (invokable.shouldInterruptOnCancel()) {
+                        executer.interrupt();
+                    }
+                    try {
+                        cancellationFuture.get(
+                                /* wait with timeout to still release the resources if taskCancellationTimeout is not configured */
+                                180000L, TimeUnit.MILLISECONDS);
```

Comment:
I didn't use the constant because it can be overriden in the configuration. 
But there is no reason actually to not check both values and wait for whatever is set:
```
taskCancellationTimeout > 0
        ? taskCancellationTimeout
        : TaskManagerOptions.TASK_CANCELLATION_TIMEOUT
                .defaultValue(),
```
(I'll update the PR accordingly)